### PR TITLE
Add naive cache for provers

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Before:
```C++
[       OK ] GPUTTensorCoreTest.FusionAmpereMatmul_CUDA (4571 ms)
[       OK ] NVFuserTest.FusionMagicSchedulerBatchNormalization_CUDA (2089 ms)
[       OK ] GpuViewTest.FusionReshapeReductionShmoo (17323 ms)
```

After:
```C++
[       OK ] GPUTTensorCoreTest.FusionAmpereMatmul_CUDA (3151 ms)
[       OK ] NVFuserTest.FusionMagicSchedulerBatchNormalization_CUDA (2044 ms)
[       OK ] GpuViewTest.FusionReshapeReductionShmoo (16693 ms)
```

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/expr_simplifier.cpp [hunk 1]
```diff
@@ -1527,77 +1542,99 @@ bool hasCompatibleSign(Val* x, Val* y, const Context& context) {
...
+      CACHE_AND_RETURN(true);
     }
   }
-  return false;
+  CACHE_AND_RETURN(false);
 }
 
+#undef CACHE_AND_RETURN
+
+#define CACHE_AND_RETURN(value)                                   \
```
Comment (jacobhinkle) [inline csrc/expr_simplifier.cpp:1586]: It might be clearer to rename these, e.g. `CACHE_AND_RETURN_LT` and ` CACHE_AND_RETURN_LE`.

**Review by jacobhinkle — APPROVED** [review]
This is nice! It's a much simpler alternative to #1974, which aims to reuse proofs across Contexts.

File: csrc/expr_simplifier.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/expr_simplifier.cpp]: changed

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
There is something wrong with Duo and I can not start the CI. Let me wait for a few hours and retry.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**